### ISSUE TITLE
[REVIEW]: umami: A Python package for Earth surface dynamics objective function construction

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1776]
**Submitting author:** @kbarnhart (<a href="http://orcid.org/0000-0001-5682-455X">Katherine Barnhart</a>)
**Repository:** <a href="https://github.com/TerrainBento/umami" target ="_blank">https://github.com/TerrainBento/umami</a>
**Version:** v1.0.1
**Editor:** @meg-simula
**Reviewer:** @sgrieve, @tristan-salles
**Archive:** <a href="https://doi.org/10.5281/zenodo.3521085" target="_blank">10.5281/zenodo.3521085</a>

## Status

[![status](https://joss.theoj.org/papers/1dea51d02def0ff3ef5108eb2a7af3a4/status.svg)](https://joss.theoj.org/papers/1dea51d02def0ff3ef5108eb2a7af3a4)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/1dea51d02def0ff3ef5108eb2a7af3a4"><img src="https://joss.theoj.org/papers/1dea51d02def0ff3ef5108eb2a7af3a4/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/1dea51d02def0ff3ef5108eb2a7af3a4/status.svg)](https://joss.theoj.org/papers/1dea51d02def0ff3ef5108eb2a7af3a4)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/TerrainBento/umami" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@sgrieve & @tristan-salles, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @meg-simula know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/TerrainBento/umami">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@kbarnhart) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/TerrainBento/umami">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@kbarnhart) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @sgrieve, @tristan-salles it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01776/joss.01776/10.21105.joss.01776.pdf)

**Comment by meg-simula** [comment]
@tristan-salles @sgrieve Thanks again for agreeing to review! Reviewer checklists have been generated for each you above - please take a look and dig in at will.

**Comment by tristan-salles** [comment]
I would like to thank the authors for this contribution, it has been a very enjoyable experience to explore this software and I am pleased to recommend publication, pending the minor issues highlighted below being resolved.

## Version

I noticed that the corresponding version for **umami** is still missing (from `versioneer.py` it is Version: 0.18). It would be good to clarify this, it is especially important for the `pip install` process as mentioned below.

## Installation

I was able to install the code via `conda` and from the `source code` following the instructions but I encountered an issue when using `pip` / `pip3`. It returned the following error message:

```bash
> pip install umami
 Collecting umami
  ERROR: Could not find a version that satisfies the requirement umami (from versions: none)
ERROR: No matching distribution found for umami
```

## Functionality

Working my way through the binder examples illustrate the functionality claims made in the documentation and the paper. More broadly I believe that Umami will be a valuable tool for researchers across the geosciences. 

> Potentially the authors can also provide in their readthedocs documentation an additional page on the required steps to use it with other packages than terrainbento and the Landlab Toolkit...

## Documentation

Going through the documentation I was really pleased with the details and quality of what has been produced. I also ❤️ the logo! The dependencies are listed in the `environment-dev.yml` file and are handled via both `conda` and `pip` managers making it really user-friendly. The core functionality (API) is also documented to a good level in Umami readthedocs website.

## Example usage

I went through the four part series of introductory notebooks and was impressed by the quality of the examples (especially Part 3 & 4). I think this is a great tool to calculate objective functions or objective function components for models that run fast.

On Binder I noted that when launching the first notebook, the first cell of the *Part 1: Introduction to Umami and the `Metric` Class* contains a previous output that will need to be cleaned before building the binder container:
```python
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-6ae86a672d39> in <module>
      4 from io import StringIO
      5 from landlab import RasterModelGrid, imshow_grid
----> 6 from umami import Metric
      7 from umami.calculations import aggregate

ModuleNotFoundError: No module named 'umami'
``` 

## Paper

The paper clearly outlines the motivation for the development of this package. It reads clearly and follows JOSS guidelines.

In conclusion, I am delighted to have had the opportunity to review this paper, and have been happy with the quality of the software, documentation and examples. I look forward to use this package with my own software!

**Comment by kbarnhart** [comment]
@tristan-salles thanks for your review. Based on it I made four Issues in the repo and will address them. Doing fieldwork tomorrow, so will be later this week/early next week.

**Comment by sgrieve** [comment]
Firstly I would like to thank the author for this contribution. I have been very impressed by the quality of the software presented here, as well as it's potential utility to a wide range of users across disciplines. I am happy to recommend publication, pending the resolution of the **minor** issues outlined below.

I would like to also note that the way this software has been developed, documented, tested and supported is of an exceptionally high standard. I would be happy to use this codebase as an exemplar of good practice in my teaching.

# Installation

I had no problems installing `umami` using pip on MacOS 10.14.6, inside a clean virtualenv running python 3.6.3. The install docs were clear, all dependences are listed in the environment file, with comments explaining where they are used.

# Functionality

The functionality of the software is clearly outlined in the series of notebooks, and these provide an excellent starting point for anyone wanting to use `umami` in their research. All of the claims of the functionality of the software are valid, and supported by clear documentation and examples.

# Documentation

The documentation is of a very high standard, with very detailed API docs, higher level documentation on the readthedocs site and detailed jupyter notebooks outlining the usage of the software in a range of scenarios.

# Community guidelines

There is a clear invitation to contribute to the software, although no detail on the preferred method of or type of contributions is given (eg opening issues, editing docs, PRs, etc). The code of conduct is excellent and I will be learning from this example of good practice for my next project.

# Tests

It is clear that the tests have been diligently designed to identify potential errors in the code. CI is set up on both Travis and Appveyor, ensuring the widest possible range of environments are being tested. Reading the tests, they are well written, and self-documenting, with good use of fixtures and other pytest and numpy features.

# Paper

The paper is excellently written and sets out the general problem being addressed and the wider context of the discipline. The second section of the paper neatly presents the software, its functionality, and its novelty. The description of the multiple output formats, and crucially, the rationale behind the choice of formats is great.

I noticed that one of the in text references is inconsistent with the others, with `(B. M. Adams et al., 2019)` including initials where other in text references only use surnames.

**Comment by kbarnhart** [comment]
@whedon generate pdf from branch barnhark/fix_biblio_reference

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch barnhark/fix_biblio_reference. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01776/joss.01776/10.21105.joss.01776.pdf)

**Comment by kbarnhart** [comment]
@whedon generate pdf from branch barnhark/fix_biblio_reference

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch barnhark/fix_biblio_reference. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01776/joss.01776/10.21105.joss.01776.pdf)

**Comment by kbarnhart** [comment]
@whedon generate pdf from branch barnhark/fix_biblio_reference

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch barnhark/fix_biblio_reference. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01776/joss.01776/10.21105.joss.01776.pdf)

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01776/joss.01776/10.21105.joss.01776.pdf)

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01776/joss.01776/10.21105.joss.01776.pdf)

**Comment by arfon** [comment]
@kbarnhart - I've upgraded Pandoc (from `v2.0.5` -> `v2.7.3`) which seems to have fixed the issue with the `Adams et al.` citation string / cc https://github.com/TerrainBento/umami/issues/14#issuecomment-541239203

**Comment by meg-simula** [comment]
@sgrieve and @tristan-salles Thanks for completing the reviews, much appreciated.
@kbarnhart Looks like you are well underway handling the feedback. Let me know if you have further questions or when you are ready to proceed further.

**Comment by kbarnhart** [comment]
Sounds good @meg-simula, I'm going to give @tristan-salles a little more time to see if he has any further recommendations to address [Issue #8](https://github.com/TerrainBento/umami/issues/8). Once I close that, I'll make the 1.0 release (which will address the [last remaining issue](https://github.com/TerrainBento/umami/issues/6)), and I'll let you know that I'm ready to proceed. 

Thanks all for constructive comments!

**Comment by tristan-salles** [comment]
Hi @meg-simula & @kbarnhart, I have just reviewed the issue #8 and it looks really good and I am happy to recommend this publication.

**Comment by kbarnhart** [comment]
@meg-simula, I've now resolved all the issues related to the reviews and made the v1.0.0 release. 

Let me know what you need from me to move forward. Thanks!

**Comment by meg-simula** [comment]
@kbarnhart Thanks, will do!

**Comment by meg-simula** [comment]
@sgrieve Are you satisfied with the response to your review? If yes, do you recommend accepting this submission?

**Comment by sgrieve** [comment]
Yes, I wholeheartedly recommend acceptance. 🚀

**Comment by kbarnhart** [comment]
thanks for your feedback @sgrieve and @tristan-salles and to @meg-simula for handling this! 🎉 

the DOI for the archive on Zenodo is https://doi.org/10.5281/zenodo.3489794

**Comment by meg-simula** [comment]
Great, thanks @sgrieve and @tristan-salles for your reviews, and @kbarnhart for responding to the suggestions!

**Comment by meg-simula** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01776/joss.01776/10.21105.joss.01776.pdf)

**Comment by meg-simula** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.5194/esurf-5-21-2017 is OK
- 10.5194/gmd-12-1267-2019 is OK
- 10.5194/esurf-7-87-2019 is OK
- 10.5194/esurf-5-369-2017 is OK
- 10.1002/2013JF002981 is OK
- 10.5194/esurf-2-1-2014 is OK
- 10.1016/j.envsoft.2009.12.002 is OK
- 10.1002/esp.3302 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by meg-simula** [comment]
@kbarnhart Could you please do an additional spell-check of the paper? (I assume "hight of mountain ranges" should be "height of ..." e.g.)

**Comment by meg-simula** [comment]
Also, the author order of the Zenodo archive is different from that of the paper, could you please provide a consistent ordering?

**Comment by kbarnhart** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by kbarnhart** [comment]
@meg-simula 
- [x] I spell checked the document, fixed the typo you noticed, and found one other. 
- [x] Incremented the version number.
- [x] Made a Zenodo archive with the same title and author order as the JOSS manuscript (rather than the auto-created Zenodo archive that comes from linking Zenodo with Github). 

The new version is: v1.0.1
The new Zenodo archive is: https://doi.org/10.5281/zenodo.3521085

Thank you!

**Comment by meg-simula** [comment]
@kbarnhart Awesome! Looking good to go now.

**Comment by meg-simula** [comment]
@whedon set 10.21105/joss.01776 as archive

**Comment by meg-simula** [comment]
@whedon set v1.0.1 as version

**Comment by whedon** [comment]
OK. v1.0.1 is the version.

**Comment by meg-simula** [comment]
@openjournals/joss-eics This paper is ready to go.

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
No archive DOI set. Exiting...

**Comment by danielskatz** [comment]
@whedon set 10.21105/joss.01776 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.21105/joss.01776" target="_blank">10.21105/joss.01776</a> is the archive.

**Comment by labarba** [comment]
@whedon set 10.5281/zenodo.3521085 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3521085" target="_blank">10.5281/zenodo.3521085</a> is the archive.

**Comment by danielskatz** [comment]
thanks @labarba

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.5194/esurf-5-21-2017 is OK
- 10.5194/gmd-12-1267-2019 is OK
- 10.5194/esurf-7-87-2019 is OK
- 10.5194/esurf-5-369-2017 is OK
- 10.1002/2013JF002981 is OK
- 10.5194/esurf-2-1-2014 is OK
- 10.1016/j.envsoft.2009.12.002 is OK
- 10.1002/esp.3302 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1061

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1061, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
👋 @kbarnhart - please merge the few minor changes in https://github.com/TerrainBento/umami/pull/19

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.5194/esurf-5-21-2017 is OK
- 10.5194/gmd-12-1267-2019 is OK
- 10.5194/esurf-7-87-2019 is OK
- 10.5194/esurf-5-369-2017 is OK
- 10.1002/2013JF002981 is OK
- 10.5194/esurf-2-1-2014 is OK
- 10.1016/j.envsoft.2009.12.002 is OK
- 10.1002/esp.3302 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1062

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1062, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1189168633009840128) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1063
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01776](https://doi.org/10.21105/joss.01776)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by danielskatz** [comment]
Thanks to @sgrieve, @tristan-salles for reviewing and to @meg-simula for editing!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01776/status.svg)](https://doi.org/10.21105/joss.01776)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01776">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01776/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01776/status.svg
   :target: https://doi.org/10.21105/joss.01776
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01776/status.svg)](https://doi.org/10.21105/joss.01776)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
